pkg/distributor/distributor.go

```diff
@@ -1460,32 +1460,40 @@ func (d *Distributor) labelValuesCardinality(ctx context.Context, labelNames []m
 		return nil, err
 	}
 
-	// Make sure we get a successful response from all the ingesters
-	replicationSet.MaxErrors = 0
-	replicationSet.MaxUnavailableZones = 0
+	// If we have a single zone, we require all ingesters to respond.
+	if replicationSet.ZoneCount() == 1 {
+		replicationSet.MaxErrors = 0
+	}
 
 	cardinalityConcurrentMap := &labelValuesCardinalityConcurrentMap{
-		cardinalityMap: map[string]map[string]uint64{},
+		cardinalityMapByZone: map[string]map[string]map[string]uint64{},
```

Comment:
i'm really good at spotting all the maps which aren't preallocated 
```suggestion
		cardinalityMapByZone: make(map[string]map[string]map[string]uint64, len(labelNames)),
```